### ISSUE TITLE
Allow pulling from multiple branches

### ISSUE THREAD

**Issue by strk** [WeblateOrg/weblate#7090]
### Describe the problem

In our setup we have a branch name (`master`) configured as a symbolic reference to another branch name (`main`) to allow developers to call their branch however they like. Commits pushed to one branch will of course be visible in both branches.

The problem is that the *webhook* (Gitea, in this case) payload sent to Weblate contains a a "ref" member which points to whatever branch name the developer pushed to (example:   `"ref": "refs/heads/main"`). Weblate then tries to *match* the branch name to the name specified in the configuration for components and unless the referenced head is found it will NOT update anything.

Basically what happens is that changes pushed to the configured branch (say: `main`) result in Weblate strings being updated while changes pushed to the alias (`master`) result in a `No matching repositories found!` failure.

### Describe the solution you'd like

I'd like to be able to specify multiple branch names for the `source repository branch` so that matches are found whatever alias changes are pushes  to.

**Comment by strk** [comment]
To be more of general use there could be a setting requesting Weblate to *scan* the source branch upon receiving a webhook invocation, no matter what the payload of the webhook references. Although I still think being able to specify multiple branches is better. For comparison, on the Gitea side, you can specify MULTIPLE branches for which a hook is invoked...

**Comment by nijel** [comment]
Updating a repository can be an expensive operation, that's why Weblate does it only when the branch matches. There can be repos with thousands of feature branches and we definitely do not want Weblate to fetch the repo on every branch update.

On the other side, using symrefs for this seems to be discouraged by Git developers, see https://gitlab.com/gitlab-org/gitlab-foss/-/issues/20370#note_104539427

**Comment by strk** [comment]
Another alternative could be extending Gitea to optionally automatically replace the "ref" field in the webhook payload when the pushed ref is a symbolic one ("canonicalize reference")

**Comment by strk** [comment]
but I still think the simplest solution would be for Weblate to allow doing some aliasing (for example: consider branch named "main" to be the same as branch named "master")

**Comment by strk** [comment]
Where is the code performing the match ?

**Comment by nijel** [comment]
There is per service helper to extract the data from the request:

https://github.com/WeblateOrg/weblate/blob/ffd50ebc839e5d11bbbbb5d8486bf3b7d35d1095/weblate/trans/views/hooks.py#L391-L403

This data is later used to filter components:

https://github.com/WeblateOrg/weblate/blob/ffd50ebc839e5d11bbbbb5d8486bf3b7d35d1095/weblate/trans/views/hooks.py#L204-L205

I still don't think handling this belongs to Weblate. This is an obscure use case discouraged by Git developers. From Weblate POV, it makes to sense to update `main` branch when `master` is updated. Every tool working with branches will have such issue. 

If you think symrefs are the way to go, it should be fixed in Gitea to understand that and to trigger webhooks on the real branch only.

**Comment by strk** [comment]
I'm afraid reporting the "real branch" would go against the use case you reported in https://gitlab.com/gitlab-org/gitlab-foss/-/issues/20370#note_104539427 - where the users actually WANT to call their branch in the "symbolic" way, to avoid having to update all integration (is my guess)

**Comment by nijel** [comment]
That issue was closed without fixing and that's pretty much my opinion as well:

>  I don't think symrefs are intended to be used in the way proposed in this issued.
